eodag/plugins/download/aws.py

```diff
@@ -287,6 +281,132 @@ def download(
+            ignore_assets,
+            product,
+        )
+
+        total_size = sum([p.size for p in unique_product_chunks])
+
+        # download
+        progress_callback.reset(total=total_size)
+        try:
+            for product_chunk in unique_product_chunks:
+                try:
+                    chunk_rel_path = self.get_chunk_dest_path(
+                        product,
+                        product_chunk,
+                        build_safe=build_safe,
+                    )
+                except NotAvailableError as e:
+                    # out of SAFE format chunk
+                    logger.warning(e)
+                    continue
+                chunk_abs_path = os.path.join(product_local_path, chunk_rel_path)
+                chunk_abs_path_dir = os.path.dirname(chunk_abs_path)
+                if not os.path.isdir(chunk_abs_path_dir):
+                    os.makedirs(chunk_abs_path_dir)
+
+                if not os.path.isfile(chunk_abs_path):
+                    product_chunk.Bucket().download_file(
+                        product_chunk.key,
+                        chunk_abs_path,
+                        ExtraArgs=getattr(s3_objects, "_params", {}),
+                        Callback=progress_callback,
+                    )
+
+        except AuthenticationError as e:
+            logger.warning("Unexpected error: %s" % e)
+        except ClientError as e:
+            err = e.response["Error"]
+            auth_messages = [
+                "AccessDenied",
+                "InvalidAccessKeyId",
+                "SignatureDoesNotMatch",
+            ]
+            if err["Code"] in auth_messages and "key" in err["Message"].lower():
+                raise AuthenticationError(
+                    "HTTP error {} returned\n{}: {}\nPlease check your credentials for {}".format(
+                        e.response["ResponseMetadata"]["HTTPStatusCode"],
+                        err["Code"],
+                        err["Message"],
+                        self.provider,
+                    )
+                )
+            logger.warning("Unexpected error: %s" % e)
+
+        # finalize safe product
+        if build_safe and "S2_MSI" in product.product_type:
+            self.finalize_s2_safe_product(product_local_path)
+        # flatten directory structure
+        elif flatten_top_dirs:
+            flatten_top_directories(product_local_path)
+
+        if build_safe:
+            self.check_manifest_file_list(product_local_path)
+
+        if asset_filter is None:
+            # save hash/record file
+            with open(record_filename, "w") as fh:
+                fh.write(product.remote_location)
+            logger.debug("Download recorded in %s", record_filename)
+
+            product.location = path_to_uri(product_local_path)
+
+        return product_local_path
+
+    def _download_preparation(
+        self, product: EOProduct, progress_callback: ProgressCallback, **kwargs
```

Comment:
- missing docstring
- missing **kwargs type

Reply:
all docstrings and types should be fixed now